src/LogicItLab/Salesforce/MapperBundle/WsdlValidator.php

```diff
@@ -0,0 +1,163 @@
+     * @param string $baseProjectDir
+     * @param string $modelDirPath
+     * @param string $wsdlPath
+     * @codeCoverageIgnore
+     */
+    public function __construct(AnnotationReader $annotationReader, string $baseProjectDir, string $modelDirPath, string $wsdlPath, string $namespacePrefix = "")
+    {
+        $this->annotationReader = $annotationReader;
+        $this->baseProjectDir = $baseProjectDir;
+        $this->modelDirPath = $modelDirPath;
+        $this->wsdlPath = $wsdlPath;
+        $this->namespacePrefix = $namespacePrefix;
+    }
+
+    public function validate(): array
+    {
+        $missingFields = [];
+
+        foreach ($this->getAllClassAnnotations() as $annotation) {
+            $objectName = $annotation['object']->name;
+            $fieldNames = $this->getFieldNames($annotation['fields']);
+
+            if (!$objectName || !$fieldNames) {
+                continue;
+            }
+
+            foreach ($fieldNames as $fieldName) {
+                if ($this->hasField($objectName, $fieldName) === false) {
+                    $missingFields[$objectName][] = $fieldName;
+                }
+            }
+        }
+
+        return $missingFields;
+    }
+
+    private function getObject(string $objectName): ?string
+    {
+        preg_match("/<complexType name=\"$objectName\">([\w\W]*)<\/complexType>/", $this->getWsdlContents(), $matches);
+
+        return $matches[0] ?? null;
+    }
+
+    private function hasField(string $objectName, string $fieldName): bool
+    {
+        return strpos($this->getObject($objectName), sprintf('<element name="%s"', $fieldName)) !== false;
+    }
+
+    private function getWsdlContents(): string
+    {
+        if (!$this->wsdlContents) {
+            $this->wsdlContents = file_get_contents($this->wsdlPath);
+        }
+
+        return $this->wsdlContents;
+    }
+
+    private function getAllClassAnnotations(): array
+    {
+        return array_map(function ($className) {
+            return $this->annotationReader->getSalesforceProperties($className);
+        }, $this->getAllClassNames());
+    }
+
+    private function getAllClassNames(): array
+    {
+        $classNames = [];
+        $AllFiles = Finder::create()->files()->in($this->baseProjectDir . $this->modelDirPath)->name('*.php');
+        foreach ($AllFiles as $file) {
+            $classNames[] = $this->getClassNameFromFile($file);
+        }
+
+        return $this->filterNonExistentClasses($classNames);
+    }
+
+    private function getClassNameFromFile($file): string
+    {
+        $realPath = $file->getRealpath();
+        $fileName = str_replace($this->baseProjectDir, '', $realPath);
+        $className = str_replace('.php', '', $fileName);
+
+        if (strpos($this->baseProjectDir, 'test') !== false) {
+            $className = "Tests/$className";
+        }
+
+        if ($this->namespacePrefix) {
+            $className = "$this->namespacePrefix/$className";
```

Comment:
Introducing a new variable like this almost always means the implementation is faulty 😃
`baseProjectDir` is always `/some_project/src`
`modelDir` can be `Comsave/Webservice` or `tests/../wherever-you-load-your-stubs-from`
Then you don't need those if weird statements. You have the full file path, not sure why you'd need a prefix if the path always matches the namespace structure.

Reply:
I introduce it because I could see that voip-api project as a prefix behind all class names that is 'App', the root folder I think. In the other projects there is not, first folder name in a namespace is whatever is the first folder after src

Reply:
Sounds good. Good catch.

Reply:
Make it by default null then as non-required.